## **Grouping Data**

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set display options for better viewing

pd.set_option('display.width', 1000)
pd.set_option('display.max_columns', None)  # Limit columns
pd.set_option('display.max_rows', 50)      # Limit rows

In [3]:
# Creating shortcut for data file
path = r'/Users/anjanpakhrin/Documents/Instacart Basket Analysis/'

In [4]:
# Create path to data file --> DataFrame was named after chapter number: "ords_prods_merge_4-7.pkl"

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4-7.pkl'))

In [5]:
# Checking output

ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                             product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day
0   2539329        1             1                  2                  8                     NaN         True         196                  1          0                                     Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders
1   2539329        1             1                  2                  8                     NaN         True       14084                  2          0  Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders
2   2539329        1             1                  2                  8                     NaN         True       12427                  3          0                      Original Beef Jerky        23             19     4.4  Low-range product  Regularly busy  Regularly busy        Average orders
3   2539329        1             1                  2                  8                     NaN         True       26088                  4          0               Aged White Cheddar Popcorn        23             19     4.7  Low-range product  Regularly busy  Regularly busy        Average orders
4   2539329        1             1                  2                  8                     NaN         True       26405                  5          0         XL Pick-A-Size Paper Towel Rolls        54             17     1.0  Low-range product  Regularly busy  Regularly busy        Average orders

#### Creating a subset of first one million entries **"df"**


In [6]:
# Create subset "df"

df = ords_prods_merge[:1000000]

In [7]:
# Check ouptput for "df"

df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                             product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day
0   2539329        1             1                  2                  8                     NaN         True         196                  1          0                                     Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders
1   2539329        1             1                  2                  8                     NaN         True       14084                  2          0  Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders
2   2539329        1             1                  2                  8                     NaN         True       12427                  3          0                      Original Beef Jerky        23             19     4.4  Low-range product  Regularly busy  Regularly busy        Average orders
3   2539329        1             1                  2                  8                     NaN         True       26088                  4          0               Aged White Cheddar Popcorn        23             19     4.7  Low-range product  Regularly busy  Regularly busy        Average orders
4   2539329        1             1                  2                  8                     NaN         True       26405                  5          0         XL Pick-A-Size Paper Towel Rolls        54             17     1.0  Low-range product  Regularly busy  Regularly busy        Average orders

In [8]:
# Check dimensions of "df"

df.shape

(1000000, 18)

In [9]:
df.groupby('product_name')

Message in above **OUTPUT** means the command is incomplete. The *df.groupby()* can be completed by adding actions such as:
- **.sum():** Add up the numbers in each group
- **.mean():** Find the average for each group
- **.count():** Counts how many items are in each group
- **.size:** Similar to **count**, but simpler

In [10]:
# Checking the frequency of each group

df.groupby('product_name').count()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered  aisle_id  department_id  prices  price_range_loc  busiest_day  busiest_days  busiest_period_of_day
product_name                                                                                                                                                                                                                                                                                                
#2 Coffee Filters                                         28       28            28                 28                 28                      25           28          28                 28         28        28             28      28               28           28            28                     28
#4 Natural Brown Coffee Filters                            2        2             2                  2                  2                       2            2           2                  2          2         2              2       2                2            2             2                      2
& Go! Hazelnut Spread + Pretzel Sticks                    25       25            25                 25                 25                      25           25          25                 25         25        25             25      25               25           25            25                     25
0 Calorie Acai Raspberry Water Beverage                    5        5             5                  5                  5                       5            5           5                  5          5         5              5       5                5            5             5                      5
0 Calorie Fuji Apple Pear Water Beverage                   6        6             6                  6                  6                       6            6           6                  6          6         6              6       6                6            6             6                      6
...                                                      ...      ...           ...                ...                ...                     ...          ...         ...                ...        ...       ...            ...     ...              ...          ...           ...                    ...
with Xylitol Minty Sweet Twist 18 Sticks Sugar ...         1        1             1                  1                  1                       1            1           1                  1          1         1              1       1                1            1             1                      1
with Xylitol Original Flavor 18 Sticks Sugar Fr...        34       34            34                 34                 34                      34           34          34                 34         34        34             34      34               34           34            34                     34
with Xylitol Unwrapped Original Flavor 50 Stick...         2        2             2                  2                  2                       2            2           2                  2          2         2              2       2                2            2             2                      2
with Xylitol Unwrapped Spearmint 50 Sticks Suga...         4        4             4                  4                  4                       3            4           4                  4          4         4              4       4                4            4             4                      4
with Xylitol Watermelon Twist 18 Sticks Sugar F...         1        1             1                  1                  1                       1            1           1                  1          1         1              1       1                1            1             1                      1

[31240 rows x 17 columns]

#### **Grouping Data with Pandas** - *"df.groupby('product_name')"*
- Split the data into groups based on some criteria
- Apply a function to each group separately
- Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe

## **Aggregating Data with agg()**

#### - Performing a Single Aggregation

In [11]:
# Grouping by "department_id" and calculating mean values for "order_number"

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                14.800024
2                17.091743
3                17.913544
4                17.893092
5                15.214270
6                15.382135
7                17.694027
8                16.458105
9                15.957363
10               20.091818
11               16.482026
12               15.615061
13               16.484023
14               17.524632
15               15.691875
16               18.014071
17               16.150593
18               19.602850
19               17.631340
20               17.138607
21               21.956893

**This shows the average number of orders per user for each department ID.**

In [12]:
# Calculating "aagregate" without using "agg()"

df.groupby('department_id')['order_number'].mean()

department_id
1     14.800024
2     17.091743
3     17.913544
4     17.893092
5     15.214270
6     15.382135
7     17.694027
8     16.458105
9     15.957363
10    20.091818
11    16.482026
12    15.615061
13    16.484023
14    17.524632
15    15.691875
16    18.014071
17    16.150593
18    19.602850
19    17.631340
20    17.138607
21    21.956893
Name: order_number, dtype: float64

#### **DOT NOTATION VS. SQUARE BRACKETS**
**DOT NOTATION:**
- **Main job:** Doing actions & asking for info (a multi-tool)
- **Best for:** Using fuctions *df.means()* and attributes *df.shape*

**SQUARE BRACKETS:**
- **Main job:** Selecting data (a specialist)
- **Best for:** Picking columns *df['column']* and rows *df[0:5]*
- Clearer and faster

In [13]:
# Calculate "mean" with "DOT NOTATION"

df.groupby('department_id').order_number.mean()

department_id
1     14.800024
2     17.091743
3     17.913544
4     17.893092
5     15.214270
6     15.382135
7     17.694027
8     16.458105
9     15.957363
10    20.091818
11    16.482026
12    15.615061
13    16.484023
14    17.524632
15    15.691875
16    18.014071
17    16.150593
18    19.602850
19    17.631340
20    17.138607
21    21.956893
Name: order_number, dtype: float64

In [14]:
# Calculate "mean" with "SQUARE BRACKETS"

df.groupby('department_id')['order_number'].mean()

department_id
1     14.800024
2     17.091743
3     17.913544
4     17.893092
5     15.214270
6     15.382135
7     17.694027
8     16.458105
9     15.957363
10    20.091818
11    16.482026
12    15.615061
13    16.484023
14    17.524632
15    15.691875
16    18.014071
17    16.150593
18    19.602850
19    17.631340
20    17.138607
21    21.956893
Name: order_number, dtype: float64

#### - Performing a Multiple Aggregations

In [15]:
# Calculation "mean", "min", and "max" of "order_number"

df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                14.800024   1  99
2                17.091743   1  98
3                17.913544   1  99
4                17.893092   1  99
5                15.214270   1  99
6                15.382135   1  99
7                17.694027   1  99
8                16.458105   1  91
9                15.957363   1  99
10               20.091818   1  99
11               16.482026   1  99
12               15.615061   1  99
13               16.484023   1  99
14               17.524632   1  99
15               15.691875   1  99
16               18.014071   1  99
17               16.150593   1  99
18               19.602850   1  99
19               17.631340   1  99
20               17.138607   1  99
21               21.956893   1  97

## **Aggregating Data with transform()**

#### Creating a flag for customers under following **criteria**:
- If **max_orders > 40**, then **"Loyal customer"**
- If **10 < max_orders <= 40**, then **"Regular customer"**
- If **max_orders <= 10**, then **"New customer"**

1. Split the data into groups based on "user_id" column.
2. Apply **transform()** function on the "order_number" column to generate the maximum orders for each user.
3. Create a new column, "max_order," to put the results of aggregation

In [16]:
# Create new column "max_order"
# Group by "user_id"
# Transform "order_number" by applying "max()" function from NumPy library
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform('max')

In [17]:
# Checking output

ords_prods_merge.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                             product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order
0   2539329        1             1                  2                  8                     NaN         True         196                  1          0                                     Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders         10
1   2539329        1             1                  2                  8                     NaN         True       14084                  2          0  Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders         10
2   2539329        1             1                  2                  8                     NaN         True       12427                  3          0                      Original Beef Jerky        23             19     4.4  Low-range product  Regularly busy  Regularly busy        Average orders         10
3   2539329        1             1                  2                  8                     NaN         True       26088                  4          0               Aged White Cheddar Popcorn        23             19     4.7  Low-range product  Regularly busy  Regularly busy        Average orders         10
4   2539329        1             1                  2                  8                     NaN         True       26405                  5          0         XL Pick-A-Size Paper Towel Rolls        54             17     1.0  Low-range product  Regularly busy  Regularly busy        Average orders         10
5   2398795        1             2                  3                  7                    15.0        False         196                  1          1                                     Soda        77              7     9.0  Mid-range product  Regularly busy    Slowest days        Average orders         10
6   2398795        1             2                  3                  7                    15.0        False       10258                  2          0                               Pistachios       117             19     3.0  Low-range product  Regularly busy    Slowest days        Average orders         10
7   2398795        1             2                  3                  7                    15.0        False       12427                  3          1                      Original Beef Jerky        23             19     4.4  Low-range product  Regularly busy    Slowest days        Average orders         10
8   2398795        1             2                  3                  7                    15.0        False       13176                  4          0                   Bag of Organic Bananas        24              4    10.3  Mid-range product  Regularly busy    Slowest days        Average orders         10
9   2398795        1             2                  3                  7                    15.0        False       26088                  5          1               Aged White Cheddar Popcorn        23             19     4.7  Low-range product  Regularly busy    Slowest days        Average orders         10

## Deriving Columns with loc()

#### Creating new flagging column in dataframe "ords_prods_merge"

In [18]:
# If max_order > 40, then "Loyal customer"
ords_prods_merge.loc[ords_prods_merge['max_order'] > 15, 'loyalty_flag'] = 'Loyal customer'

# If 10 < max_orders <= 40, then "Regular customer"
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order']> 10), 'loyalty_flag'] = 'Regular customer'

# If max_orders <= 10, then "New customer"
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 10, 'loyalty_flag')] = 'New customer'

In [19]:
# Checking output for new column
ords_prods_merge.groupby('loyalty_flag').head(2)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                             product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag
0      2539329        1             1                  2                  8                     NaN         True         196                  1          0                                     Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer
1      2539329        1             1                  2                  8                     NaN         True       14084                  2          0  Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer
59     2168274        2             1                  2                 11                     NaN         True       32792                  1          0           Chipotle Beef & Pork Realstick        23             19     5.2  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer
60     2168274        2             1                  2                 11                     NaN         True       47766                  2          0                          Organic Avocado        24              4     6.3  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer
2474    361493       27             1                  3                  9                     NaN         True       25718                  1          0                         Cold Brew Coffee        26              7    10.4  Mid-range product  Regularly busy    Slowest days           Most orders         81    Loyal customer
2475   1662354       27             2                  2                 17                     6.0        False       30776                  1          0           Organic Raw Kombucha Gingerade        31              7    13.4  Mid-range product  Regularly busy  Regularly busy        Average orders         81    Loyal customer

In [20]:
# Checking frequency for each category
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

**Majority of customers fall into the category "Regular customer"**

In [21]:
# Return the only the rows of 'user_id', 'loyalty_flag', and 'order_number'
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             1
2         1      New customer             1
3         1      New customer             1
4         1      New customer             1
..      ...               ...           ...
55        1      New customer            10
56        1      New customer            10
57        1      New customer            10
58        1      New customer            10
59        2  Regular customer             1

[60 rows x 3 columns]

#### **Cleaning Outliers**

In [22]:
# Mean value for "prices"
float(ords_prods_merge['prices'].mean())

11.980225563049316

In [23]:
# Median for "prices"
float(ords_prods_merge['prices'].median())

7.400000095367432

In [24]:
# Max for "prices"
float(ords_prods_merge['prices'].max())

99999.0

#### Identifying **OUTLIERS**
* A value of 100 for prices can be defined as threshold assuming that an item in a supermarket would cost in average $100.
* Anything above this threshold could be considered as an outlier.

In [25]:
# Running the threshold check - Outlier observation

ords_prods_merge.loc[ords_prods_merge['prices'] > 100]

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                      product_name  aisle_id  department_id   prices     price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag
1576        912404       17            12                  2                 14                     5.0        False       21553                  5          0  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product  Regularly busy  Regularly busy           Most orders         40  Regular customer
1638        603376       17            22                  6                 16                     4.0        False       21553                  3          1  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product  Regularly busy  Regularly busy           Most orders         40  Regular customer
16522      3264360      135             2                  2                 21                    13.0        False       21553                  6          0  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product  Regularly busy  Regularly busy        Average orders          4      New customer
16528       892534      135             3                  0                  8                    12.0        False       21553                  3          1  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product     Busiest day    Busiest days        Average orders          4      New customer
53672       229704      342             8                  1                 19                    30.0        False       21553                  9          0  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product  Regularly busy    Busiest days        Average orders         16  Regular customer
...            ...      ...           ...                ...                ...                     ...          ...         ...                ...        ...                               ...       ...            ...      ...                 ...             ...             ...                   ...        ...               ...
32320511   3172853   205650            18                  1                  9                     7.0        False       21553                 17          1  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product  Regularly busy    Busiest days           Most orders         25  Regular customer
32347717   2504315   205818             3                  5                 15                     3.0        False       21553                 13          0  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product  Regularly busy  Regularly busy           Most orders         25  Regular customer
32347727   1108388   205818             5                  4                  5                     1.0        False       21553                  5          1  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product      Least busy    Slowest days         Fewest orders         25  Regular customer
32380527   1916142   206049             1                  2                 17                     NaN         True       21553                  2          0  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product  Regularly busy  Regularly busy        Average orders          5      New customer
32380551    379732   206049             4                  1                 14                     5.0        False       21553                  4          1  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  High-range product  Regularly busy    Busiest days           Most orders          5      New customer

[5127 rows x 20 columns]

#### **Observations:**
* There are 5,127 rows with outlier observations (prices > $100).
* Prices like $14,900.00 or $99,9999 per item don't make sense in terms of the other values in column.
* Values > 100, therefore, defined as outliers and marked as missing (NaN).
#### Turning values > 100 into NaNs

In [26]:
# Turning outliers into NaNs

ords_prods_merge.loc[ords_prods_merge['prices'] > 100, 'prices'] = np.nan

In [27]:
# Descriptive statistics

ords_prods_merge['prices'].describe()

count    3.239973e+07
mean     7.790994e+00
std      4.100926e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64